Java writer: missing parentheses for a conditional used as a condition and for a cast of a negated value

### Bug

The Java source writer renders two expression shapes without the parentheses Java needs, producing source that compiles to a different expression or does not compile.

**1. A conditional expression as the condition of another conditional**

```java
new ExpressionDef.IfElse(new ExpressionDef.IfElse(a, b, ExpressionDef.constant(false)), ExpressionDef.constant(true), c)
```

renders as `a ? b : false ? true : c`, which Java parses right-associatively as `a ? b : (false ? true : c)` — a different expression. Expected `(a ? b : false) ? true : c`. The bytecode writer is correct.

**2. A cast of a negated literal or a unary operation**

```java
ExpressionDef.constant(-1L).cast(TypeDef.OBJECT)            // or
ExpressionDef.constant(1L).math(MathUnaryOperation.OpType.NEGATE).cast(TypeDef.OBJECT)
```

renders as `(Object) -1l`, which javac parses as the subtraction `Object - 1l`: `error: cannot find symbol: variable Object`. Expected `(Object) (-1l)`. Only reference-type casts are affected, `(long) -1` is fine.

### Root cause

`JavaPoetSourceGenerator`: `case ExpressionDef.IfElse` renders `condition.condition()` with `renderExpression`, not `renderExpressionWithParentheses`; `requiresCastOperandParentheses` does not include `MathUnaryOperation` nor a negative numeric `Constant`.

### Environment

micronaut-sourcegen 2.1.0; found in micronaut-jakarta-el compiling `a && b || c` and `cond ? x : -1`.

--- a/sourcegen-generator-java/src/main/java/io/micronaut/sourcegen/JavaPoetSourceGenerator.java
+++ b/sourcegen-generator-java/src/main/java/io/micronaut/sourcegen/JavaPoetSourceGenerator.java
@@ -1027,8 +1027,13 @@ public sealed class JavaPoetSourceGenerator implements SourceGenerator permits G
                 );
             }
             case ExpressionDef.IfElse condition -> {
+                CodeBlock conditionBlock = renderExpression(objectDef, methodDef, scope, condition.condition());
+                if (unwrapCasts(condition.condition()) instanceof ExpressionDef.IfElse) {
+                    // `?:` is right-associative, a conditional used as a condition needs parentheses
+                    conditionBlock = addParentheses(conditionBlock);
+                }
                 return CodeBlock.concat(
-                    renderExpression(objectDef, methodDef, scope, condition.condition()),
+                    conditionBlock,
                     CodeBlock.of(" ? "),
                     renderExpression(objectDef, methodDef, scope, condition.ifExpression()),
                     CodeBlock.of(" : "),
@@ -1263,8 +1268,17 @@ public sealed class JavaPoetSourceGenerator implements SourceGenerator permits G
         return expressionDef instanceof ExpressionDef.ConditionExpressionDef
             || expressionDef instanceof ExpressionDef.IfElse
             || expressionDef instanceof ExpressionDef.MathBinaryOperation
+            || expressionDef instanceof ExpressionDef.MathUnaryOperation
             || expressionDef instanceof ExpressionDef.StringConcatenation
-            || expressionDef instanceof ExpressionDef.Switch;
+            || expressionDef instanceof ExpressionDef.Switch
+            || isNegativeNumericConstant(expressionDef);
+    }
+
+    private static boolean isNegativeNumericConstant(ExpressionDef expressionDef) {
+        // `(Object) -1` would parse as a subtraction of the variable `Object`
+        return expressionDef instanceof ExpressionDef.Constant constant
+            && constant.value() instanceof Number number
+            && number.toString().startsWith("-");
     }
 
     private static boolean requiresMethodCallTargetParentheses(ExpressionDef expressionDef) {

--- a/sourcegen-generator-java/src/test/java/io/micronaut/sourcegen/javapoet/write/ExpressionWriteTest.java
+++ b/sourcegen-generator-java/src/test/java/io/micronaut/sourcegen/javapoet/write/ExpressionWriteTest.java
@@ -1221,6 +1221,58 @@ public class Example {
     }
 
     @Test
+    void returnIfElseWithIfElseConditionParentheses() throws IOException {
+        ExpressionDef expression = new ExpressionDef.IfElse(
+            new ExpressionDef.IfElse(ExpressionDef.constant(true), ExpressionDef.constant(false), ExpressionDef.constant(false)),
+            ExpressionDef.constant("yes"),
+            ExpressionDef.constant("no")
+        );
+        String result = writeMethodWithExpression(expression);
+
+        assertEquals("(true ? false : false) ? \"yes\" : \"no\"", result);
+    }
+
+    @Test
+    void returnIfElseWithIfElseElseBranchWithoutParentheses() throws IOException {
+        ExpressionDef expression = new ExpressionDef.IfElse(
+            ExpressionDef.constant(true),
+            ExpressionDef.constant("a"),
+            new ExpressionDef.IfElse(ExpressionDef.constant(false), ExpressionDef.constant("b"), ExpressionDef.constant("c"))
+        );
+        String result = writeMethodWithExpression(expression);
+
+        assertEquals("true ? \"a\" : false ? \"b\" : \"c\"", result);
+    }
+
+    @Test
+    void returnCastedNegativeConstantWithParentheses() throws IOException {
+        ExpressionDef castedExpression = ExpressionDef.constant(-1L)
+            .cast(TypeDef.OBJECT);
+        String result = writeMethodWithExpression(castedExpression);
+
+        assertEquals("(Object) (-1l)", result);
+    }
+
+    @Test
+    void returnCastedNegatedValueWithParentheses() throws IOException {
+        ExpressionDef castedExpression = ExpressionDef.constant(1L)
+            .math(NEGATE)
+            .cast(TypeDef.OBJECT);
+        String result = writeMethodWithExpression(castedExpression);
+
+        assertEquals("(Object) (-1l)", result);
+    }
+
+    @Test
+    void returnPrimitiveCastedNegativeConstantWithParentheses() throws IOException {
+        ExpressionDef castedExpression = ExpressionDef.constant(-1)
+            .cast(TypeDef.Primitive.LONG);
+        String result = writeMethodWithExpression(castedExpression);
+
+        assertEquals("(long) (-1)", result);
+    }
+
+    @Test
     void returnCastedMathOperationWithParentheses() throws IOException {
         ExpressionDef castedExpression = ExpressionDef.constant(1)
             .math(ADDITION, ExpressionDef.constant(2))

